Added "Principles" to explain the key technical principles behind Fastify (#4852)
fastify/fastify: docs/Reference/Principles.md

# Technical Principles

Every decision in the Fastify framework and its official plugins is guided by the following technical principles:

1. “Zero” overhead in production;
2. “Good” developer experience;
3. Works great for small & big projects alike;
4. Easy to migrate to microservices (or even serverless) and back;
5. Security & data validation;
6. If something could be a plugin, it likely should
7. Easily testable;
8. Do not monkeypatch core;
9. Semantic versioning & Long Term Support;
10. Specification adherence.

## "Zero" Overhead in Production

Fastify aims to implement its features by adding as minimal overhead to your application as possible.
This is usually delivered by implementing fast algorithms and data structures, as well as JavaScript-specific
features.

> Given that JavaScript does not offer zero-overhead data structures, ™his principle is at add with providing

Reviewer: ```suggestion
Given that JavaScript does not offer zero-overhead data structures, this principle is at add with providing
```

# Technical Principles

Every decision in the Fastify framework and its official plugins is guided by
the following technical principles:

> 1. “Zero” overhead in production;
> 2. “Good” developer experience;
> 3. Works great for small & big projects alike;
> 4. Easy to migrate to microservices (or even serverless) and back;
> 5. Security & data validation;
> 6. If something could be a plugin, it likely should
> 7. Easily testable;
> 8. Do not monkeypatch core;
> 9. Semantic versioning & Long Term Support;
> 10. Specification adherence.

Reviewer 2: ```suggestion
1. “Zero” overhead in production
2. “Good” developer experience
3. Works great for small & big projects alike
4. Easy to migrate to microservices (or even serverless) and back
5. Security & data validation
6. If something could be a plugin, it likely should be
7. Easily testable
8. Do not monkeypatch core
9. Semantic versioning & Long Term Support
10. Specification adherence
```

## "Zero" Overhead in Production

Fastify aims to implement its features by adding as minimal overhead to your
application as possible.
This is usually delivered by implementing fast algorithms and data structures,
as well as JavaScript-specific features.

Given that JavaScript does not offer zero-overhead data structures, this principle
is at add with providing a great developer experience and providing more features,

> as usually those cost some overhead.

Reviewer 2: Bit lost with this sentence as "is at add" doesn't make sense?

Author: ```suggestion
Given that JavaScript does not offer zero-overhead data structures, this principle
is at odd with providing a great developer experience and providing more features,
as usually those cost some overhead.
```

Author: add -> odd

## "Good" Developer Experience

Fastify aims to provide the best developer experience at the performance point
it is operating.

> It provides a great out-of-the-box experience that it is flexible enough to be
> adapted to a various level of situations.

Reviewer 2: ```suggestion
It provides a great out-of-the-box experience that is flexible enough to be
adapted to a variety of situations.
```

As an example, this means that binary addons are forbidden because most JavaScript
developers would not
have access to a compiler.

## Works great for small and big projects alike

We recognize that most applications start small and become more complex over time.
Fastify aims to grow with
the complexity of your application, providing advanced features to structure
your codebase.

## Easy to migrate to microservices (or even serverless) and back

How you deploy your routes should not matter. The framework should "just work".

## Security and Data Validation

Your web framework is the first point of contact with untrusted data, and it

> needs to act as the first line of defence for your system.

Reviewer 2: ```suggestion
needs to act as the first line of defense for your system.
```
Even though it hurts my soul, the rest of the documentation uses American English.

American English: Defense
British English: Defence

## If something could be a plugin, it likely should

We recognize that there are an infinite amount of use cases for an HTTP framework
for Node.js. Catering to them in a single module would make the codebase unmaintainable.

> Thefore we provide hooks and options to allow you to customize the framework

Reviewer 2: ```suggestion
Therefore we provide hooks and options to allow you to customize the framework
```

as you please.

## Easily testable

> Testing Fastify applications should be a first class concern.

Reviewer 2: ```suggestion
Testing Fastify applications should be a first-class concern.
```
